---
title: "Metabolomics report"
author: Nicholas Sunderland
output:
  pdf_document:
    toc: true
    toc_depth: 2
    number_sections: true
    # includes:
    #   in_header: latex_styles.sty 
urlcolor: blue
params:
  workflow_dag: "x"
  associations_volcanos: "x"
  intermodel_correlations: "x"
  lmm_time_assoc_bbs_metabolon: "x"
  lmm_time_assoc_bbs_nightingale: "x"
  lmm_time_assoc_direct_metabolon: "x"
  lmm_time_assoc_direct_nightingale: "x"
  volcano_bbs_direct: "x"
  volcano_bmi_bbs_direct: "x"
  corr_bbs_direct: "x"
  corr_bmi_bbs_direct: "x"
  bbs_direct_rep: "x"
  bbs_direct_mr_rep: "x"
  bbs_mr: "x"
  bbs_cluster_scree: "x"
  bbs_clusters: "x"
  bbs_cluster_stability: "x"
  bbs_drug_target_mr: "x"
  bbs_cis_pqtl_corr_hist: "x"
  bbs_cis_pqtl_nsnp: "x"
  bbs_cis_pqtl_mr: "x"
  bbs_cis_pqtl_corr: "x"
  hf_mr_qq: "x"
---

\newpage

# Acknowledgements
Substantial work has already been accomplished on this project by members of the Timpson group. The initial metabolite modelling and comparative analysis was completed by Dr Maddy Smith, as described in her thesis *Obesity and health: Insight from characterizing the metabolomic signatures of weight-loss interventions*. The metabolomics quality control pipelines were developed by Dr David Hughes and Dr Laura Corbin and published as the [metaboprep](https://pubmed.ncbi.nlm.nih.gov/35134881/) R-package.

\newpage


# Rationale  
Adverse adiposity, indexed through body mass index (BMI), is an established risk factor of a wide variety of diseases. Higher BMI is 
strongly associated with the development of heart failure (HF), in particular HF with preserved ejection fraction. Weight loss 
has a multitude of benefits but is difficult to achieve and maintain. There is a growing number of efficacious pharmaceutical and surgical strategies by which to induce clinically meaningful weight loss. However some, such as bariatric surgery, are invasive and not all people who would benefit from weight loss are eligible. 

The mechanisms by which these interventions translate into improved outcomes is incompletely understood. For example, it is not known whether weight loss in general is the driving factor, or whether there are effects specific to the intervention. This is important as there is a clinical choice to be made regarding which interventions we offer individuals living with obesity. In addition, greater biological understanding of the metabolic pathways involved in weight loss will aid in further drug development efforts to either induce weight loss or mimic its beneficial effect.  


## The circulating metabololomic signature of weight loss?  
High throughput metabolomic studies offer the opportunity to get a read out on the circulating metabolic profile of the obese state 
and how this is perturbed from the norm. However, any differences found between obese and normal weight, or less overweight, profiles may not necessarily point to relevant metabolic pathways, depending on the situation. They could be related to intervention, medication use, bystander phenomena, sample processing or measurement noise.  

Repeated measures, pre and post weight loss intervention, may help narrow the search to metabolites strongly related to 
BMI. However, changes may be intervention-specific and not necessarily related to BMI change per-se. By analysing 
how circulating profiles change across multiple types of weight loss intervention the search can be narrowed down further 
to metabolites more specifically related to weight loss. Even then, the profile change will not be entirely specific as
there are common strands to undergoing any weight loss intervention, such as engagement in other positive lifestyle changes.  

Mendelian randomisation (MR) studies utilising genetic variants strongly associated with BMI as the exposure, and metabolic genome-wide association analyses (mGWAS) as the outcome, may add another layer of evidence corroborating metabolites as strongly related to BMI.  


## Heart failure  
BMI has been identified through clinical trials and MR studies as a major causal risk factor for the development of HF. Emerging trail evidence supports the role of weight loss to improve outcomes in individuals with established HF and preserved ejection fraction (HFpEF), however there is concern regarding weight loss strategies in HF with reduced ejection fraction (HFrEF) as a result of conflicted data from several small randomisation controlled trials (RCTs) as well as observational data. 

The mechanisms by which elevated BMI induces different diseased states need not be the same. By looking at the overlap between 
metabolite profiles of weight loss and that of HF, either through direct metabolomic profiling of HF patients or through 
instrumental variable analyses, may shed light on the components of adverse adiposity most important in the development of 
HF and therefore targets for prevention and treatment of the condition. 

\newpage


# Aims
1. The first aim is to identify the metabolites altered by two weight loss interventions (bariatric surgery and calorie restriction) - this is a replication of Maddie's work. 

2. Secondly, I aimed to compare the effect of weight loss intervention on metabolites, with the effect of (genetically-predicted) BMI and metabolite levels in the general population (two sample MR analysis).  

3. Thirdly, I attempt to assess the predicted metabolomic profile of several weight loss drugs, through drug-target MR, and compare this to that seen with weight loss intervention.

\newpage


# Data  

## By-Band-Sleeve  
By-Band-Sleeve (BBS) was an open-label RCT comparing the effectiveness of several types of bariatric surgery: gastric bypass, gastric band and sleeve gastrectomy (registration number: ISRCTN00786323). 1341 patients were recruited. Metabolomic profiling on the Metabolon mass spectrometry platform at baseline and follow up post-surgery lead to a total of 1434 samples for analysis after removal of participants who withdrew consent. 1331 metabolites were measured, of which 1253 remained after quality control. Profiling on the NMR Nightingale platform lead to a total of 1410 samples with 250 metabolites after similar QC and exclusions.  

## DiRECT  
The Diabetes Remission Clinical Trial (DiRECT) was a cluster-RCT assessing the implementation of a weight management programme delivered in primary care (registration number: ISRCTN03267836). Participants were randomised to either calorie restriction (825-853 kcal/day) or usual care. Sampling was conducted at baseline and 12-months. 574 samples (302 individuals) with metabolic profiling of 1276 metabolites on the Metabolon platform are available. In addition, 227 metabolites from the Nightingale NMR platform were also measured. 


## mGWAS   
Metabolite genome-wide association statistics for the Nightingale NMR platform metabolites have been previously published by [Karjalainen et al.](https://pubmed.ncbi.nlm.nih.gov/38448586/) They profiled 233 metabolites in 136,016 participants from 33 cohorts, not including the UK Biobank.

\newpage


# Analysis  
## Setup  
To reproduce the analyses the appropriate environment variables (raw data paths) need to be set in the `.env.` file. These are not published to GitLab so you would need to contact [Nick Sunderland](nicholas.sunderland@bristol.ac.uk) to populate these correctly. An example `.env_example` file is provided with the environment variable names to be populated. 

The appropriate analysis packages need to be configured in a conda environment, details on how to set this up are located in the `wt1_wp1_036_bmi_hf_metabolomics` GitLab repository README.


## Work-flow
The below figure presents a rule graph of the analysis pipeline which has been implemented in Snakemake. Each rule takes input files from source, or previous rules, through analysis R-scripts located in the `/scripts` folder of the `wt1_wp1_036_bmi_hf_metabolomics` Timpson GitLab. The rules are defined in the `Snakemake` file in at the top level of the repository.  

The terminal command to run the pipeline is: 
```{bash, eval=FALSE, echo=TRUE}
cd path_to_repository/wt1_wp1_036_bmi_hf_metabolomics
conda activate wt1_wp1_036_bmi_hf_metabolomics
snakemake -c2 all
```

```{r, echo=FALSE, out.width="100%"}
knitr::include_graphics(params$workflow_dag)
```

\newpage


# Models and data types  
## Data types
Whilst we can work with the underlying raw abundance metabolomic data, it is often best to use several data transformations. The Metabolon platform is semi-quantitative and so I will use the rank inverse normal transformed version of the data. The Nightgale NMR platform data is quantitative and so working with the raw data, or Z-score normalised data such that the associations can retain meaningful units.

Data types:

* `raw` - raw untransformed data.
* `raw_z` - z-score normalised data.
* `raw_filtered` - filtered for metabolites with > 750 samples (BBS only).
* `rnt` - rank inverse normal transformed.
* `rnt_filtered` - filtered, as before, `rnt` data (BBS only).
* `pa` - presence-absence coding of metabolites with >50 missing and >50 present samples.

## Models  
There are a number of possible models to assess the effect of intervention on the metabolite levels. For the BBS study, where there is no control arm, we are mainly interested in the estimate for time post intervention. For the DiRECT study there is an intervention and control arm and so the estimate of interest is the interaction term of time with treatment-arm. Additionally, I will look at the effect of BMI, which is clearly highly linked to time post weight loss intervention.

### Linear
A linear model of the effect of time post-intervention on the metabolite level. Does not incorporate repeated measures. 

$$ Metabolite \sim timepoint + (\pm tp*treat_{BBS}) + sex + age + site + study + time~in~freezer$$

### Linear mixed model  
A linear mixed model with incorporates the repeated measures within individuals with the inclusion of the random effects term. 

$$ Metabolite \sim timepoint + (\pm tp*treat_{BBS}) + sex + age + site + study + time~in~freezer + (1|id) $$

### Linear mixed model with BMI. 
A linear mixed model with BMI instead of time, looking at the effect of BMI change on metabolite levels. BMI was scaled to $\mu=0$, $\sigma=1$, making the estimate reflect change in metabolite per standard deviation change in BMI. This makes comparison with the MR results easier.

$$ Metabolite \sim scaled~BMI + age + sex + site + study + time~in~freezer + (1|id) $$

### Linear mixed model with time and BMI  
A linear mixed model with BMI and time, looking at the effect of BMI change on metabolite levels as well as the effect from the intervention. In reality I don't think this will work well as the time post intervention and BMI change are linked and there will be no reason to think that the model partitions the effect sensibly.  

$$ Metabolite \sim timepoint + (\pm tp*treat_{BBS}) + BMI + age + sex + site + study + time~in~freezer + (1|id) $$

### Linear mixed model: time, baseline BMI, and delta-BMI  
This model contains the change in BMI from timepoint-0 to timepoint-N, i.e. for timepoint-0 the delta will be zero and for the end-timepoint the delta will be the change in weight over follow up. The baseline BMI term accounts for between-subjects BMI differences. I don't think this is great, it has similar problems to the previous model.

$$ Metabolite \sim tp + (\pm tp*treat_{BBS}) + \Delta BMI + baseline~BMI + age + sex + site + study + time~in~freezer + (1|id) $$
\newpage


# Metabolite associations
The below plot present the associations for the different models run using the different data types. The associations are all for the effect of timepoint post intervention. 

```{r, echo=FALSE, out.width="100%"}
knitr::include_graphics(params$associations_volcanos)
```


## Model association correlations
Next I looked at the differences between the models. Below are the correlations between the association statistics obtained using the different models (only the BBS RNT Metabolon data presented). We see that the timepoint betas for the linear time model are roughly the same as for the linear mixed model, although the error is smaller with the linear mixed model owing to the incorporation of the repeated measures with the random effect `id` term.  

The betas for BMI in the linear mixed model with BMI instead of time also correlate well, negatively as expected, indicating that the majority of the time after intervention effect can be explained through change in BMI. 

```{r, echo=FALSE, out.width="100%"}
knitr::include_graphics(params$intermodel_correlations)
```


## Top metabolite associations
Below are the tables of the top metabolite associations in the BBS and DiRECT studies from both profiling platforms.

```{r, echo=FALSE}
library(knitr)
library(kableExtra)
library(data.table)
library(ggplot2)

files <- list(
  "BBS"    = params$lmm_time_assoc_bbs_metabolon, 
  "BBS"    = params$lmm_time_assoc_bbs_nightingale,
  "DiRECT" = params$lmm_time_assoc_direct_metabolon,
  "DiRECT" = params$lmm_time_assoc_direct_nightingale
)


dat <- lapply(files, fread) |> rbindlist(idcol = "study")
tbl_timepoint <- dat[grepl("timepoint", term)][order(p.value), .(study, chemical_name, estimate, p.value)]
tbl_timepoint[, p.value := formatC(p.value, format = "e", digits = 2)]


kable(head(tbl_timepoint[study=="BBS", .(chemical_name, estimate, p.value)], 20),
      caption = "BBS - Top Associations with Timepoint (end)",
      digits = 4,
      col.names = c("Chemical Name", "Estimate", "P-value")) |>
  kable_styling(latex_options = c("striped"), full_width = FALSE) |> 
  column_spec(column = 1, width = "10cm") |> 
  column_spec(column = 2:3, width = "2cm")


kable(head(tbl_timepoint[study=="DiRECT", .(chemical_name, estimate, p.value)], 20),
      caption = "DiRECT - Top Associations with Timepoint (end)",
      digits = 4,
      col.names = c("Chemical Name", "Estimate", "P-value")) |>
  kable_styling(latex_options = c("striped"), full_width = FALSE) |> 
  column_spec(column = 1, width = "10cm") |> 
  column_spec(column = 2:3, width = "2cm")
```

\newpage


# Study comparison
Next I look at the replication of metabolite associations across the two studies. There are several ways in which consistency of association could be defined:  

1. consistent direction of effect  
2. consistent direction of effect and nominal significance P value < 0.05  
3. consistent direction of effect and FDR P value < 0.05 or Holm adjustment  
4. consistent direction of effect and significance at a power adjusted P value  

Options 3 and 4 are most reasonable.  


## Associations - timepoint
The plot below presents the associations (timepoint/intervention betas) for BBS and DiRECT for the two measurement platforms (MS - RNT data, NMR - raw Z-normalised data).
```{r, echo=FALSE, out.width="100%"}
knitr::include_graphics(params$volcano_bbs_direct)
```

## Correlation of time point effect
The plot below presents the correlation between timepoint effect estimates from BBS and DiRECT for the two measurement platforms. The regression line is calculated using the `mcr::mcreg` Deming regression method for comparison of two measurements, incorporating the error of both. The metabolite names with the greatest orthogonal residuals are plotted.
```{r, echo=FALSE, out.width="100%"}
knitr::include_graphics(params$corr_bbs_direct)
```

## Associations - BMI
The plot below presents the BMI associations for BBS and DiRECT for the two measurement platforms (MS - RNT data, NMR - raw Z-normalised data).
```{r, echo=FALSE, out.width="100%"}
knitr::include_graphics(params$volcano_bmi_bbs_direct)
```

## Correlation of BMI effect
The plot below presents the correlation between BMI effect estimates from BBS and DiRECT for the two measurement platforms. Potentially better correlation compared to timepoint associations. I need to find a pseudo-R2 for the Deming regression to quantify ... `1 - (rss / tss)`?. 
```{r, echo=FALSE, out.width="100%"}
knitr::include_graphics(params$corr_bmi_bbs_direct)
```


## Replication  
### Power-adjusted replication  
Metabolites BMI betas that replicate in DiRECT after accounting for differences in power ([see here](https://explodecomputer.github.io/lab-book/posts/2024-02-27-prop-replication/)).
```{r, echo=FALSE, out.width="100%"}
knitr::include_graphics(params$bbs_direct_rep)
```


\newpage


# BMI-Metabolite Mendelian Randomisation  
For the Nightingale platform there are mGWAS summary statistics publicly available - [Karjalainen et al.](https://pubmed.ncbi.nlm.nih.gov/38448586/). I have run the MR of exposure: `BMI` on outcome: `metabolite` levels. David Hughes has also sent me the MR results for the Metabolon platform and so I intend to incorporate these soon. 

## BMI effect on metabolites vs. weight loss effect
We see the correlation between metabolites given genetically predicted BMI and that seen in the BBS study.
```{r, echo=FALSE, out.width="100%"}
knitr::include_graphics(params$bbs_mr)
```

## BMI effect on metabolites vs. BBS vs. DiRECT
Joining the the MR estimates to the results from the BMI models in the two trials and assessing replication with a power adjusted replication across studies, and a nominal MR P value < 0.05 we can plot the 3 lines of evidence. 
```{r, echo=FALSE, out.width="100%"}
knitr::include_graphics(params$bbs_direct_mr_rep)
```

\newpage


# Replication within clusters
This next section explored the idea of defining replication within clusters. The idea being that many metabolites are similar and that replication may fail for a specific metabolite, however it may be that the general cluster of metabolites does show the same direction of change - i.e. there may be a strong association with weight loss of one other member of the cluster.  

The initial question is how well do the metabolites cluster, what is the cluster stability?

## Clustering
I undertook this clustering process: 

1. take the baseline metabolite levels from the BBS study.
2. take metabolites with <5% missingness.
3. median imputation of the remaining missing values.
4. principal component analysis.
5. parallel analysis to identify the number (`Nsig`) of PCs carrying greater than random noise.
6. extract the loadings of `Nsig` PCs.
7. perform dynamic clustering (`dynamicTreeCut::cutreeDynamic`) to assign features to clusters based on their loadings pattern.
8. assess cluster stability (`fpc::clusterboot`).

### Significant PCs
58 PCs carrying greater than noise information. 
```{r, echo=FALSE, out.width="100%"}
knitr::include_graphics(params$bbs_cluster_scree)
```

### Clusters
Clusters of PC loadings
```{r, echo=FALSE, out.width="100%"}
knitr::include_graphics(params$bbs_clusters)
```

### Cluster stability
We see that the majority of clusters have poor stability (great uncertainty that a metabolite relates to others in that cluster, below the red line) and none have high stability (above blue line).
```{r, echo=FALSE, out.width="100%"}
knitr::include_graphics(params$bbs_cluster_stability)
```

### Conclusion 
At least with this clustering method I do no think assessing replication by cluster is workable, as it is not clear which cluster a metabolite should belong to. 

\newpage


# Weight loss drug-target metabolomic profiles
I next aimed to briefly explore the predicted metabolomic profile with some common weight loss therapeutics, namely GLP1R agonists and SLGT2 inhibitors. The instruments themselves need more thought and discussion and the whole endevour probably has some methodological holes that need considering. However, I took the gene regions (+/- 250kb) for GLP1R and SLC5A2 and attempted to instrument them using GWAS of HbA1c and also BMI.  

For GLP1R, there is no variant associating with BMI at P < 5e-8 in the region +/- 250kb. Others have attempted this same instrument and ended up extending the region to 1Mb which includes a variant upstream of the gene - [Burgess et al.](https://www.medrxiv.org/content/10.1101/2025.02.12.25322150v1) I therefore tried this same extended region.  

## Cis-MR GLP1R & SLC5A2
There is some positive correction instrumenting these targets. The GLP1R plot looks odd but I don't think I am instrumenting with BMI at all well. 
```{r, echo=FALSE, out.width="100%"}
knitr::include_graphics(params$bbs_drug_target_mr)
```

\newpage


# Search for 'weight loss' proteins - druggable targets?
Tom had the idea of asking with proteins are responsible for the metabolite perturbations (i.e. druggable targets). We could use cis-pQTLs and perform the cis-MR of these proteins on all of the metabolites.  

* exposure: protein levels (cis pQTLs within 250kb of gene)
* outcome: metabolite levels

For each protein we would then be able to correlate the MR result against the association of the metabolite with weight loss intervention. Proteins with high correlation might indicate a biological role in weight loss / BMI. This might reveal novel proteins.  

Here are some examples of the UKB-PPP pQTL data run through that type of pipeline:

### Protein instrument size (independent cis-pQTLs)
```{r, echo=FALSE, out.width="100%"}
knitr::include_graphics(params$bbs_cis_pqtl_nsnp)
```

### Protein instrument effect vs. weight-loss effect correlation
```{r, echo=FALSE, out.width="100%"}
knitr::include_graphics(params$bbs_cis_pqtl_corr_hist)
```

### Top proteins with similar metabolomic profiles to weight loss
```{r, echo=FALSE, out.width="100%"}
knitr::include_graphics(params$bbs_cis_pqtl_mr)
```

\newpage


# Metabolite - Heart Failure MR
```{r, echo=FALSE, out.width="100%"}
knitr::include_graphics(params$hf_mr_qq)
```

## Metabolites as effectors of disease
Another idea Tom and I discussed was taking the metabolites that show greatest change with weight loss and running the cis-MR of these onto disease, when the canonical protein is known. 

\newpage

# Future directions 

## Instrumenting a metabolic profiles  
Laura and I discussed instrumenting profiles of metabolites and the previous work done in the group looking at GWASs of metabolite PCs, and using these as exposures. It's a bit complicated, abstract and difficult to immediately see the analytical route forward here as well as the utility, but it could be interesting to discuss.  

## Applying weight loss profiles to populations  
This seemed interesting at the time, but now I'm not sure. Essentially I looked at the profiles of NMR (Nightingale) metabolites in the UK Biobank population by their BMI category assigned at UKBB enrollment. Below is the plot of the first, second, third and fourth principal components. The general direction for underweight and normal weight is opposite to high BMI categories. 

![UKBB NMR PCs](/Users/xx20081/git/bmi_metabolomics/output/figures/ukbb_metabolomics_pca.png)

I then experimented with 'applying' the weight loss associations to individuals with high BMI, simulating the average effect of weight loss intervention. For example, if the $\beta$ for metabolite X was 1.0 and the weight loss $\beta$ was -0.3, the resulting $\beta$ would be 0.7. I then added the new data to the original PCA space, looking for a shift back towards a normal weight profile, which it does to some extent. 

![UKBB NMR PCs simulated intervention in high BMI individuals](/Users/xx20081/git/bmi_metabolomics/output/figures/ukbb_metabolomics_pca_post_bbs.png)

# End  
Look forward to discussing. 
